src/tools/rbd_mirror/Replayer.cc

```diff
@@ -0,0 +1,125 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/debug.h"
+#include "common/errno.h"
+#include "include/stringify.h"
+#include "Replayer.h"
+
+#define dout_subsys ceph_subsys_rbd_mirror
+#undef dout_prefix
+#define dout_prefix *_dout << "rbd-mirror: "
+
+using std::map;
+using std::string;
+using std::unique_ptr;
+using std::vector;
+
+namespace rbd {
+namespace mirror {
+
+Replayer::Replayer(RadosRef primary_cluster, const peer_t &peer) :
+  m_lock(stringify("rbd::mirror::Replayer ") + stringify(peer)),
+  m_peer(peer),
+  m_primary(primary_cluster)
+{
+}
+
+Replayer::~Replayer()
+{
+}
+
+int Replayer::start()
+{
+  int r = m_remote->init2(m_peer.client_name.c_str(),
+			  m_peer.cluster_name.c_str(), 0);
+  if (r < 0) {
+    derr << "error initializing remote cluster for " << m_peer
+	 << " : " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  r = m_remote->conf_read_file(nullptr);
+  if (r < 0) {
+    derr << "could not read ceph conf for " << m_peer
+	 << " : " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  r = m_remote->connect();
+  if (r < 0) {
+    derr << "error connecting to remote cluster " << m_peer
+	 << " : " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  string cluster_uuid;
+  r = m_remote->cluster_fsid(&cluster_uuid);
+  if (r < 0) {
+    derr << "error reading cluster uuid from remote cluster " << m_peer
+	 << " : " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  if (cluster_uuid != m_peer.cluster_uuid) {
+    derr << "configured cluster uuid does not match actual cluster uuid. "
+	 << "expected: " << m_peer.cluster_uuid
+	 << " observed: " << cluster_uuid << dendl;
+    return -EINVAL;
+  }
+
+  return 0;
+}
+
+void Replayer::stop()
+{
+  m_remote->shutdown();
+}
+
+void Replayer::set_sources(const map<int64_t, set<string> > &images)
```

Comment:
I was writing this assuming a push model, but can change to pull instead. I don't think there was a good reason for thinking of it as push - there's not much local sharing that would be useful. Now that I think about failover scenarios, pull does seem simpler.

For image removals, I was thinking we could treat them like [dis|en]abling mirroring on an image (and send a notify when we do that), but more and more it seems like storing the mirroring state in ``rbd_directory`` would be better than using generic image flags. I was thinking deletion would be mirrored as well.

This reminds me of another thing I don't think we've discussed:
Should ``rbd-mirror`` keep mirrored pools strictly in sync among clusters, or only operate on images with mirroring enabled? i.e., if non-mirrored images are created in a secondary cluster, should we do anything (warn, delete), or ignore them? I've been thinking the latter in both cases, but if a new mirrored image is created in the primary cluster, and the secondary cluster has one with the same name (without mirroring enabled), we have a problem. Perhaps in this kind of case we should expose a warning as part of the mirroring status, so the admin can decide what to do.

Another aspect is preventing accidental modification of mirrored images in a secondary cluster. It seems we'd want to disable writes to them at least, and perhaps only allow changing mirroring state (enabling/disabling) in the primary cluster.

we may want a way to prevent mirroring from being enabled for images in a secondary cluster.